File: easyplot/scripts/easyplot_analysis.py
"""Explicit unit summaries and inference: pandas/NumPy/SciPy, no file writes."""

from numbers import Real

import numpy as np
import pandas as pd
from scipy import stats


def _identities(values, name):
    """Keep homogeneous text/numeric identities without string coercion."""
    values = [x.item() if isinstance(x, np.generic) else x for x in values]
    kinds = set()
    for x in values:
        if isinstance(x, str) and x.strip():
            kinds.add("text")
        elif (isinstance(x, Real) and not isinstance(x, bool)
              and (isinstance(x, int) or np.isfinite(x))):
            kinds.add("numeric")
        else:
            raise ValueError(f"{name} must contain nonempty text or finite numeric identities")
    if len(kinds) != 1:
        raise ValueError(f"{name} must contain one identity type (text or numeric)")
    return values, kinds.pop()


def summarize_replicates(data, *, group, value, unit, replicate, design, missing="error"):
    """Average comparable technical readings per unit/condition, in first-seen order.

    Design is declared as independent or repeated. Identity checks precede
    exclusions. Available readings retain all-missing cells as None and never
    create absent conditions. SD is within-cell sample dispersion, not SE/CI.
    Downstream unit means have equal weight regardless of technical counts.
    """
    if not isinstance(data, pd.DataFrame) or data.empty:
        raise ValueError("data must be a nonempty DataFrame")
    columns = [group, value, unit, replicate]
    if (any(not isinstance(x, str) or not x.strip() for x in columns)
            or len(set(columns)) != 4 or not data.columns.is_unique
            or any(x not in data.columns for x in columns)):
        raise ValueError("group, value, unit and replicate must name distinct, existing, unique columns")
    if not isinstance(design, str) or design not in ("independent", "repeated"):
        raise ValueError("design must be 'independent' or 'repeated'")
    if not isinstance(missing, str) or missing not in ("error", "available"):
        raise ValueError("missing must be 'error' or 'available'")
    groups, _ = _identities(data[group], "group")
    units, _ = _identities(data[unit], "unit")
    replicates, _ = _identities(data[replicate], "replicate")
    # Native tuple keys avoid pandas' lossy inference for mixed large ints/floats.
    seen, cells, unit_groups = set(), {}, {}
    for i, (u, g, r) in enumerate(zip(units, groups, replicates)):
        key = (u, g, r)
        if key in seen:
            raise ValueError("duplicate unit/group/replicate")
        seen.add(key)
        if design == "independent" and u in unit_groups and unit_groups[u] != g:
            raise ValueError("independent design requires each unit in a single group")
        unit_groups[u] = g
        cells.setdefault((u, g), []).append(i)
    dtype = data[value].dtype
    if (not pd.api.types.is_numeric_dtype(dtype) or pd.api.types.is_bool_dtype(dtype)
            or pd.api.types.is_complex_dtype(dtype)):
        raise ValueError("value must be a real numeric column, excluding booleans")
    measurements = data[value].to_numpy(dtype=float, na_value=np.nan)
    if np.isinf(measurements).any():
        raise ValueError("value must not contain infinite measurements")
    n_missing = int(np.isnan(measurements).sum())
    if n_missing and missing == "error":
        raise ValueError("missing measurements require missing='available'")
    summary = []
    for (u, g), rows in cells.items():
        sample = measurements[rows]
        sample = sample[~np.isnan(sample)]
        with np.errstate(over="ignore", invalid="ignore"):
            mean = float(np.mean(sample)) if len(sample) else None
            sd = float(np.std(sample, ddof=1)) if len(sample) > 1 else None
        if any(x is not None and not np.isfinite(x) for x in (mean, sd)):
            raise ValueError("nonfinite replicate summary; rescale measurements")
        summary.append(dict(unit=u, group=g, value=mean, n_input=len(rows),
                            n_used=len(sample), n_missing=len(rows) - len(sample), sd=sd))
    warnings = ["Technical readings must not be counted as independent sample size (n).",
                "Within-unit SD describes technical dispersion only; biological SE/CI require unit-level inference."]
    if missing == "available":
        warnings.append("Available-readings summaries may introduce selection bias when readings are missing.")
    if any(s["n_used"] == 0 for s in summary):
        warnings.append("All-missing unit/condition rows are retained with null value and SD; absent conditions are not imputed.")
    return {
        "schema_version": "1.0.0", "design": design,
        "columns": dict(group=group, value=value, unit=unit, replicate=replicate),
        "estimator": "arithmetic mean within unit/condition", "missing_policy": missing,
        "n_input": len(data), "n_used": len(data) - n_missing, "n_missing": n_missing,
        "n_units": len(unit_groups), "n_unit_conditions": len(summary), "summary": summary,
        "warnings": warnings,
        "assumptions": [
            "Readings within each unit/condition are comparable technical repetitions.",
            "Unit/condition keys capture all relevant design dimensions; no time, dose, or other dimension is ignored.",
            "The estimand is the arithmetic mean within each unit/condition; downstream unit means receive equal weight regardless of the number of readings.",
            "Design and replicate roles are declared by the caller; hierarchical or mixed-model estimands require separate validation.",
        ],
    }


def adjust_pvalues(pvalues, *, labels, method, family, alpha=.05):
    """Adjust one explicit, complete comparison family; preserve input order."""
    if not isinstance(method, str) or method not in ("holm", "bonferroni", "BH", "BY"):
        raise ValueError("method must be 'holm', 'bonferroni', 'BH' or 'BY'")
    if not isinstance(family, str) or not family.strip():
        raise ValueError("family must be a nonempty description of the comparison family")
    if (isinstance(alpha, (bool, np.bool_)) or not isinstance(alpha, Real)
            or not 0 < alpha < 1 or not np.isfinite(alpha)):
        raise ValueError("alpha must be a finite number strictly between 0 and 1")
    sequences = (list, tuple, np.ndarray, pd.Series, pd.Index)
    if (not isinstance(pvalues, sequences) or isinstance(pvalues, np.ma.MaskedArray)
            or getattr(pvalues, "ndim", 1) != 1 or not len(pvalues)):
        raise ValueError("pvalues must be a nonempty one-dimensional sequence")
    raw = list(pvalues)
    if any(isinstance(x, (bool, np.bool_)) or not isinstance(x, Real)
           or not 0 <= x <= 1 or not np.isfinite(x) for x in raw):
        raise ValueError("pvalues must contain finite real numbers in [0, 1], without missing values")
    if (not isinstance(labels, sequences) or isinstance(labels, np.ma.MaskedArray)
            or getattr(labels, "ndim", 1) != 1 or len(labels) != len(raw)):
        raise ValueError("labels must be a one-dimensional sequence matching pvalues")
    labels = list(labels)
    if (any(not isinstance(x, str) or not x.strip() for x in labels)
            or len(set(labels)) != len(labels)):
        raise ValueError("labels must be unique nonempty strings")
    ps = np.array(raw, dtype=float)
    m = len(ps)
    if method == "bonferroni":
        adjusted = np.minimum(1., ps * m)
    elif method == "holm":
        order = np.argsort(ps, kind="stable")
        adjusted = np.empty(m)
        adjusted[order] = np.minimum(1., np.maximum.accumulate(ps[order] * np.arange(m, 0, -1)))
    else:
        adjusted = stats.false_discovery_control(ps, method=method.lower())
    dependence = ("FWER control is valid under arbitrary dependence among comparisons."
                  if method in ("holm", "bonferroni") else
                  "BH controls FDR under independence or positive regression dependence on a subset (PRDS) of the true null hypotheses."
                  if method == "BH" else
                  "BY controls FDR under arbitrary dependence among comparisons.")
    return {
        "schema_version": "1.0.0", "family": family, "method": method,
        "error_control": "FWER" if method in ("holm", "bonferroni") else "FDR",
        "alpha": float(alpha), "m": m,
        "results": [dict(comparison=label, p_raw=float(p), p_adjusted=float(q), reject=bool(q <= alpha))
                    for label, p, q in zip(labels, ps, adjusted)],
        "assumptions": ["All supplied raw p-values are valid for their null hypotheses.",
                        "The comparison family was specified by the caller; no failed or missing p-values were removed.",
                        dependence],
    }


def analyze_two_groups(data, *, group, value, unit, order, design,
                       missing="error", confidence=.95):
    """Return schema 1.0.0; effect direction is always order[1] - order[0].

    `data` is an unchanged DataFrame; column names must be distinct strings.
    Group/order and unit identities are homogeneous nonblank text or finite
    real numbers (factors/categoricals are accepted). Values must have a real
    numeric dtype, excluding booleans. No string-to-number coercion is done.

    Independent samples require globally unique units, n >= 2 and positive
    sample variance per group. Paired samples require unique unit/group rows,
    >= 2 complete ID-matched pairs and positive variance of their differences.
    Complete-case deletion is opt-in. n_input/n_used/n_removed count actual
    rows; n_pairs counts retained pairs (None for independent). removed_units
    lists excluded identities once, in first-appearance order. Warnings are
    returned as text, not emitted. No input files or global state are changed.
    """
    if not isinstance(data, pd.DataFrame) or data.empty:
        raise ValueError("data must be a nonempty DataFrame")
    columns = [group, value, unit]
    if (any(not isinstance(x, str) or not x.strip() for x in columns)
            or len(set(columns)) != 3 or not data.columns.is_unique
            or any(x not in data.columns for x in columns)):
        raise ValueError("group, value and unit must name distinct, existing, unique columns")
    if not isinstance(design, str) or design not in ("independent", "paired"):
        raise ValueError("design must be 'independent' or 'paired'")
    if not isinstance(missing, str) or missing not in ("error", "complete-case"):
        raise ValueError("missing must be 'error' or 'complete-case'")
    if (isinstance(confidence, (bool, np.bool_)) or not isinstance(confidence, Real)
            or not np.isfinite(confidence) or not 0 < confidence < 1):
        raise ValueError("confidence must be a finite number strictly between 0 and 1")
    if isinstance(order, (str, bytes, dict, set, frozenset)):
        raise ValueError("order must be an ordered sequence of exactly two distinct groups")
    try:
        order = list(order)
    except TypeError as exc:
        raise ValueError("order must contain exactly two distinct groups") from exc
    if len(order) != 2:
        raise ValueError("order must contain exactly two distinct groups")
    order, order_kind = _identities(order, "order")
    groups, group_kind = _identities(data[group], "group")
    units, _ = _identities(data[unit], "unit")
    if order[0] == order[1] or order_kind != group_kind:
        raise ValueError("order must contain two distinct groups of the group column's type")
    # Encode identities before pandas can infer a lossy float dtype. Python
    # dict equality preserves distinct large-int/float labels through pivoting.
    unit_labels = list(dict.fromkeys(units))
    unit_codes = {label: i for i, label in enumerate(unit_labels)}
    group_codes = {label: i for i, label in enumerate(order)}
    if any(label not in group_codes for label in groups):
        raise ValueError("data contains a group outside order")
    work = pd.DataFrame({"unit": [unit_codes[label] for label in units],
                         "group": [group_codes[label] for label in groups]})
    # Identity checks precede every exclusion, including missing measurements.
    keys = ["unit"] if design == "independent" else ["unit", "group"]
    if work.duplicated(keys).any():
        raise ValueError("duplicate unit" if design == "independent" else "duplicate unit/group")
    dtype = data[value].dtype
    if (not pd.api.types.is_numeric_dtype(dtype) or pd.api.types.is_bool_dtype(dtype)
            or pd.api.types.is_complex_dtype(dtype)):
        raise ValueError("value must be a real numeric column, excluding booleans")
    measurements = data[value].to_numpy(dtype=float, na_value=np.nan)
    if np.isinf(measurements).any():
        raise ValueError("value must not contain infinite measurements")
    work["value"] = measurements
    n_input = len(work)
    if design == "independent":
        absent = work["value"].isna()
        if absent.any() and missing == "error":
            raise ValueError("missing measurements require missing='complete-case'")
        removed_units = [unit_labels[i] for i in work.loc[absent, "unit"]]
        used = work.loc[~absent]
        samples = [used.loc[used["group"] == i, "value"].to_numpy() for i in range(2)]
        n_used, n_pairs = len(used), None
    else:
        wide = work.pivot(index="unit", columns="group", values="value").reindex(
            index=work["unit"].drop_duplicates(), columns=range(2))
        absent = wide.isna().any(axis=1)
        if absent.any() and missing == "error":
            raise ValueError("missing or absent pair members require missing='complete-case'")
        removed_units = [unit_labels[i] for i in wide.index[absent]]
        used = wide.loc[~absent]
        samples = [used[i].to_numpy() for i in range(2)]
        n_pairs = len(used)
        n_used = 2 * n_pairs
    if any(len(sample) < 2 for sample in samples):
        raise ValueError("at least two observations per group or two complete pairs are required")

    first, second = samples
    with np.errstate(over="ignore", invalid="ignore", divide="ignore"):
        summary = [{"group": label, "n": len(sample), "mean": float(np.mean(sample)),
                    "sd": float(np.std(sample, ddof=1)),
                    "se": float(stats.sem(sample))}
                   for label, sample in zip(order, samples)]
        variance_samples = samples if design == "independent" else [second - first]
        variances = [np.var(sample, ddof=1) for sample in variance_samples]
        if any(not np.isfinite(x) or x <= 0 for x in variances):
            raise ValueError("positive finite sample variance is required (paired: differences)")
        if any(not np.isfinite(s[key]) for s in summary for key in ("mean", "sd", "se")):
            raise ValueError("nonfinite sample summary; rescale measurements")
        # Match R t.test's numerical admissibility guard; inference stays in SciPy.
        precision_se = (np.hypot(summary[0]["se"], summary[1]["se"])
                        if design == "independent" else stats.sem(second - first))
        precision_scale = (max(abs(s["mean"]) for s in summary) if design == "independent"
                           else abs(np.mean(second - first)))
        if precision_se < 10 * np.finfo(float).eps * precision_scale:
            raise ValueError("data are essentially constant")
        if design == "independent":
            result = stats.ttest_ind(second, first, equal_var=False, alternative="two-sided")
            estimate = float(np.mean(second) - np.mean(first))
        else:
            result = stats.ttest_rel(second, first, alternative="two-sided")
            estimate = float(np.mean(second - first))
        ci = result.confidence_interval(confidence_level=float(confidence))
    if not np.isfinite([estimate, ci.low, ci.high, result.statistic, result.df, result.pvalue]).all():
        raise ValueError("nonfinite inference result; check variance and measurement scale")

    n_removed = n_input - n_used
    warnings = (["Complete-case exclusions may introduce selection bias."] if n_removed else [])
    assumptions = (["Units are independent within and between groups.",
                    "Within-group observations are approximately normal, or sample sizes justify the t approximation.",
                    "Equal population variances are not required."] if design == "independent" else
                   ["Pairs are matched by unit ID; different units are independent.",
                    "Within-unit differences are approximately normal, or the number of pairs justifies the t approximation."])
    assumptions.append("Design and distributional assumptions are supplied by the caller, not inferred from data.")
    return {
        "schema_version": "1.0.0", "design": design,
        "columns": {"group": group, "value": value, "unit": unit}, "order": order,
        "missing_policy": missing, "n_input": n_input, "n_used": n_used,
        "n_removed": n_removed, "n_pairs": n_pairs, "removed_units": removed_units,
        "summary": summary,
        "effect": {"label": "mean difference (second - first)", "estimate": estimate,
                   "ci_low": float(ci.low), "ci_high": float(ci.high), "confidence": float(confidence)},
        "test": {"method": "Welch Two Sample t-test" if design == "independent" else "Paired t-test",
                 "statistic": float(result.statistic), "df": float(result.df),
                 "p_value": float(result.pvalue), "alternative": "two-sided"},
        "warnings": warnings, "assumptions": assumptions,
    }

File: easyplot/scripts/easyplot_analysis.R
# Explicit unit summaries and inference. Base R/stats; no exports or global options.
# Schema 1.0.0 matches easyplot_analysis.py. Effect = order[2] - order[1].
# Identity columns/order: homogeneous nonblank text/factors or finite numbers.
# n_input/n_used/n_removed count actual rows; n_pairs counts retained pairs or NULL.
# removed_units is a list in first-appearance order. Input data is never modified.
.easyplot_nonblank <- function(x) nzchar(trimws(x, whitespace = "[\\h\\v]"))
.easyplot_text_scalar <- function(x) {
  is.character(x) && length(x) == 1L && is.null(dim(x)) && !is.na(x) && .easyplot_nonblank(x)
}
.easyplot_real_numeric <- function(x) is.numeric(x) && !is.complex(x) && !is.object(x) && is.null(dim(x))
.easyplot_identities <- function(x, name) {
  if (is.factor(x)) x <- as.character(x)
  if ((!is.character(x) && !.easyplot_real_numeric(x)) || !is.null(dim(x)) || anyNA(x) ||
      (is.character(x) && any(!.easyplot_nonblank(x))) ||
      (is.numeric(x) && any(!is.finite(x)))) {
    stop(paste(name, "must contain nonempty text or finite numeric identities"), call. = FALSE)
  }
  unname(x)
}

easyplot_summarize_replicates <- function(data, group, value, unit, replicate, design, missing = "error") {
  fail <- function(message) stop(message, call. = FALSE)
  if (!is.data.frame(data) || nrow(data) == 0L) fail("data must be a nonempty data.frame")
  columns <- list(group, value, unit, replicate)
  if (!all(vapply(columns, .easyplot_text_scalar, logical(1))) ||
      length(unique(unlist(columns))) != 4L || anyDuplicated(names(data)) ||
      !all(unlist(columns) %in% names(data))) {
    fail("group, value, unit and replicate must name distinct, existing, unique columns")
  }
  if (!.easyplot_text_scalar(design) || !design %in% c("independent", "repeated")) {
    fail("design must be 'independent' or 'repeated'")
  }
  if (!.easyplot_text_scalar(missing) || !missing %in% c("error", "available")) {
    fail("missing must be 'error' or 'available'")
  }
  groups <- .easyplot_identities(data[[group]], "group")
  units <- .easyplot_identities(data[[unit]], "unit")
  replicates <- .easyplot_identities(data[[replicate]], "replicate")
  # Integer codes preserve identity without pasted-label collisions or sorting.
  unit_codes <- match(units, unique(units))
  group_codes <- match(groups, unique(groups))
  keys <- data.frame(unit = unit_codes, group = group_codes,
                     replicate = match(replicates, unique(replicates)))
  if (anyDuplicated(keys)) fail("duplicate unit/group/replicate")
  cells <- !duplicated(keys[c("unit", "group")])
  if (design == "independent" && anyDuplicated(unit_codes[cells])) {
    fail("independent design requires each unit in a single group")
  }
  measurements <- data[[value]]
  if (!.easyplot_real_numeric(measurements)) fail("value must be a real numeric column, excluding booleans")
  measurements <- as.numeric(measurements)
  if (any(is.infinite(measurements))) fail("value must not contain infinite measurements")
  n_missing <- sum(is.na(measurements))
  if (n_missing > 0L && missing == "error") fail("missing measurements require missing='available'")
  # Pasting integer codes is unambiguous; split once instead of scanning per cell.
  cell_keys <- paste(unit_codes, group_codes, sep = "/")
  rows <- split(seq_along(measurements), match(cell_keys, unique(cell_keys)))
  summary <- lapply(seq_along(rows), function(i) {
    indices <- rows[[as.character(i)]]
    x <- measurements[indices]
    x <- x[!is.na(x)]
    mean_value <- if (length(x)) mean(x) else NULL
    sd <- if (length(x) > 1L) stats::sd(x) else NULL
    if (any(!is.finite(c(mean_value, sd)))) fail("nonfinite replicate summary; rescale measurements")
    list(unit = units[[indices[[1L]]]], group = groups[[indices[[1L]]]], value = mean_value,
         n_input = length(indices), n_used = length(x), n_missing = length(indices) - length(x), sd = sd)
  })
  warnings <- list("Technical readings must not be counted as independent sample size (n).",
                   "Within-unit SD describes technical dispersion only; biological SE/CI require unit-level inference.")
  if (missing == "available") {
    warnings <- c(warnings, list("Available-readings summaries may introduce selection bias when readings are missing."))
  }
  if (any(vapply(summary, function(x) x$n_used == 0L, logical(1)))) {
    warnings <- c(warnings, list("All-missing unit/condition rows are retained with null value and SD; absent conditions are not imputed."))
  }
  list(schema_version = "1.0.0", design = design,
       columns = list(group = group, value = value, unit = unit, replicate = replicate),
       estimator = "arithmetic mean within unit/condition", missing_policy = missing,
       n_input = nrow(data), n_used = nrow(data) - n_missing, n_missing = n_missing,
       n_units = length(unique(units)), n_unit_conditions = length(summary), summary = summary,
       warnings = warnings, assumptions = list(
         "Readings within each unit/condition are comparable technical repetitions.",
         "Unit/condition keys capture all relevant design dimensions; no time, dose, or other dimension is ignored.",
         "The estimand is the arithmetic mean within each unit/condition; downstream unit means receive equal weight regardless of the number of readings.",
         "Design and replicate roles are declared by the caller; hierarchical or mixed-model estimands require separate validation."))
}

easyplot_adjust_pvalues <- function(pvalues, labels, method, family, alpha = .05) {
  fail <- function(message) stop(message, call. = FALSE)
  if (!.easyplot_text_scalar(method) || !method %in% c("holm", "bonferroni", "BH", "BY")) {
    fail("method must be 'holm', 'bonferroni', 'BH' or 'BY'")
  }
  if (!.easyplot_text_scalar(family)) fail("family must be a nonempty description of the comparison family")
  if (!.easyplot_real_numeric(alpha) || length(alpha) != 1L || !is.finite(alpha) || alpha <= 0 || alpha >= 1) {
    fail("alpha must be a finite number strictly between 0 and 1")
  }
  if (!.easyplot_real_numeric(pvalues) || !length(pvalues) || any(!is.finite(pvalues)) ||
      any(pvalues < 0 | pvalues > 1)) {
    fail("pvalues must be a nonempty one-dimensional sequence of finite real numbers in [0, 1], without missing values")
  }
  if (!is.character(labels) || !is.null(dim(labels)) || length(labels) != length(pvalues) ||
      anyNA(labels) || any(!.easyplot_nonblank(labels)) || anyDuplicated(labels)) {
    fail("labels must be unique nonempty strings matching pvalues")
  }
  adjusted <- stats::p.adjust(pvalues, method = method)
  dependence <- if (method %in% c("holm", "bonferroni")) {
    "FWER control is valid under arbitrary dependence among comparisons."
  } else if (method == "BH") {
    "BH controls FDR under independence or positive regression dependence on a subset (PRDS) of the true null hypotheses."
  } else {
    "BY controls FDR under arbitrary dependence among comparisons."
  }
  list(schema_version = "1.0.0", family = family, method = method,
       error_control = if (method %in% c("holm", "bonferroni")) "FWER" else "FDR",
       alpha = unname(alpha), m = length(pvalues),
       results = lapply(seq_along(pvalues), function(i) {
         list(comparison = unname(labels[[i]]), p_raw = unname(pvalues[[i]]),
              p_adjusted = unname(adjusted[[i]]), reject = unname(adjusted[[i]] <= alpha))
       }), assumptions = list("All supplied raw p-values are valid for their null hypotheses.",
                              "The comparison family was specified by the caller; no failed or missing p-values were removed.",
                              dependence))
}

easyplot_analyze_two_groups <- function(data, group, value, unit, order, design,
                                        missing = "error", confidence = .95) {
  fail <- function(message) stop(message, call. = FALSE)
  text_scalar <- .easyplot_text_scalar
  real_numeric <- .easyplot_real_numeric
  identities <- .easyplot_identities
  if (!is.data.frame(data) || nrow(data) == 0L) fail("data must be a nonempty data.frame")
  columns <- list(group, value, unit)
  if (!all(vapply(columns, text_scalar, logical(1))) ||
      length(unique(unlist(columns))) != 3L || anyDuplicated(names(data)) ||
      !all(unlist(columns) %in% names(data))) {
    fail("group, value and unit must name distinct, existing, unique columns")
  }
  if (!text_scalar(design) || !design %in% c("independent", "paired")) {
    fail("design must be 'independent' or 'paired'")
  }
  if (!text_scalar(missing) || !missing %in% c("error", "complete-case")) {
    fail("missing must be 'error' or 'complete-case'")
  }
  if (!real_numeric(confidence) || length(confidence) != 1L ||
      !is.finite(confidence) || confidence <= 0 || confidence >= 1) {
    fail("confidence must be a finite number strictly between 0 and 1")
  }
  if (length(order) != 2L) fail("order must contain exactly two distinct groups")
  order <- identities(order, "order")
  groups <- identities(data[[group]], "group")
  units <- identities(data[[unit]], "unit")
  if (anyDuplicated(order) || is.character(order) != is.character(groups)) {
    fail("order must contain two distinct groups of the group column's type")
  }
  if (!all(groups %in% order)) fail("data contains a group outside order")
  # Validate identities before complete-case exclusion can hide duplicates.
  if (design == "independent" && anyDuplicated(units)) fail("duplicate unit")
  if (design == "paired" && anyDuplicated(data.frame(unit = units, group = groups))) {
    fail("duplicate unit/group")
  }
  measurements <- data[[value]]
  if (!real_numeric(measurements)) fail("value must be a real numeric column, excluding booleans")
  measurements <- as.numeric(measurements)
  if (any(is.infinite(measurements))) fail("value must not contain infinite measurements")
  n_input <- nrow(data)
  if (design == "independent") {
    absent <- is.na(measurements)
    if (any(absent) && missing == "error") fail("missing measurements require missing='complete-case'")
    removed_units <- as.list(units[absent])
    samples <- lapply(order, function(label) measurements[!absent & groups == label])
    n_used <- sum(!absent)
    n_pairs <- NULL
  } else {
    ids <- unique(units)
    samples <- lapply(order, function(label) {
      rows <- which(groups == label)
      measurements[rows[match(ids, units[rows])]]
    })
    absent <- is.na(samples[[1L]]) | is.na(samples[[2L]])
    if (any(absent) && missing == "error") {
      fail("missing or absent pair members require missing='complete-case'")
    }
    removed_units <- as.list(ids[absent])
    samples <- lapply(samples, function(x) x[!absent])
    n_pairs <- sum(!absent)
    n_used <- 2L * n_pairs
  }
  if (any(lengths(samples) < 2L)) {
    fail("at least two observations per group or two complete pairs are required")
  }
  first <- samples[[1L]]
  second <- samples[[2L]]
  summary <- lapply(seq_along(order), function(i) {
    x <- samples[[i]]
    list(group = order[[i]], n = length(x), mean = mean(x), sd = stats::sd(x),
         se = stats::sd(x) / sqrt(length(x)))
  })
  variance_samples <- if (design == "independent") samples else list(second - first)
  variances <- vapply(variance_samples, stats::var, numeric(1))
  if (any(!is.finite(variances) | variances <= 0)) {
    fail("positive finite sample variance is required (paired: differences)")
  }
  if (any(!is.finite(unlist(lapply(summary, function(x) x[c("mean", "sd", "se")]))))) {
    fail("nonfinite sample summary; rescale measurements")
  }
  result <- stats::t.test(second, first, paired = design == "paired", var.equal = FALSE,
                          alternative = "two.sided", conf.level = confidence)
  estimate <- if (design == "paired") mean(second - first) else mean(second) - mean(first)
  if (any(!is.finite(c(estimate, result$conf.int, result$statistic, result$parameter, result$p.value)))) {
    fail("nonfinite inference result; check variance and measurement scale")
  }
  n_removed <- n_input - n_used
  warnings <- if (n_removed > 0L) list("Complete-case exclusions may introduce selection bias.") else list()
  assumptions <- if (design == "independent") {
    list("Units are independent within and between groups.",
         "Within-group observations are approximately normal, or sample sizes justify the t approximation.",
         "Equal population variances are not required.")
  } else {
    list("Pairs are matched by unit ID; different units are independent.",
         "Within-unit differences are approximately normal, or the number of pairs justifies the t approximation.")
  }
  assumptions <- c(assumptions, list("Design and distributional assumptions are supplied by the caller, not inferred from data."))
  list(
    schema_version = "1.0.0", design = design,
    columns = list(group = group, value = value, unit = unit), order = as.list(order),
    missing_policy = missing, n_input = n_input, n_used = n_used, n_removed = n_removed,
    n_pairs = n_pairs, removed_units = removed_units, summary = summary,
    effect = list(label = "mean difference (second - first)", estimate = estimate,
                  ci_low = unname(result$conf.int[1L]), ci_high = unname(result$conf.int[2L]),
                  confidence = unname(confidence)),
    test = list(method = if (design == "independent") "Welch Two Sample t-test" else "Paired t-test",
                statistic = unname(result$statistic), df = unname(result$parameter),
                p_value = result$p.value, alternative = "two-sided"),
    warnings = warnings, assumptions = assumptions
  )
}
